core/src/main/java/org/apache/calcite/rel/rules/JoinConditionExpandIsNotDistinctFromRule.java

```diff
@@ -0,0 +1,159 @@
+    Join join = call.rel(0);
+    RexNode oldJoinCond = join.getCondition();
+
+    RexCall found =
+        RexUtil.findOperatorCall(SqlStdOperatorTable.IS_NOT_DISTINCT_FROM,
+        oldJoinCond);
+
+    if (found == null) {
+      // no longer contains isNotDistinctFromOperator
+      return;
+    }
+
+    for (RexNode op : found.getOperands()) {
+      if (op.getType().getSqlTypeName() == SqlTypeName.MAP) {
+        // map type cannot be compared
+        return;
+      }
+    }
+
+    RemoveIsNotDistinctFromRexShuttle rewriteShuttle =
+        new RemoveIsNotDistinctFromRexShuttle(
+            join.getCluster().getRexBuilder());
+
+    RelNode newJoin = join.accept(rewriteShuttle);
+    assert newJoin instanceof Join;
+    if (((Join) newJoin).getCondition().equals(join.getCondition())) {
+      return;
+    }
+    call.transformTo(newJoin);
+  }
+
+  //~ Inner Classes ----------------------------------------------------------
+
+  /** The RexShuttle use for converting 'x IS NOT DISTINCT FROM y' to
+   * '(COALESCE(x, {zero}) = COALESCE(y, {zero}))
+   * AND ((x IS NULL) = (y IS NULL))'. */
+  private static class RemoveIsNotDistinctFromRexShuttle extends RexShuttle {
+    final RexBuilder rexBuilder;
+
+    RemoveIsNotDistinctFromRexShuttle(
+        RexBuilder rexBuilder) {
+      this.rexBuilder = rexBuilder;
+    }
+
+    @Override public RexNode visitCall(RexCall call) {
+      RexNode newCall = super.visitCall(call);
+
+      if (call.getOperator()
+          == SqlStdOperatorTable.IS_NOT_DISTINCT_FROM) {
+        RexCall tmpCall = (RexCall) newCall;
+
+        RexNode operand0 = tmpCall.operands.get(0);
+        RexNode operand1 = tmpCall.operands.get(1);
+
+        RexNode operand0Zero = makeZeroLiteral(rexBuilder, operand0);
+        RexNode operand1Zero = makeZeroLiteral(rexBuilder, operand1);
+
+        RexNode coalesceCondition =
+            rexBuilder.makeCall(SqlStdOperatorTable.EQUALS,
+            rexBuilder.makeCall(SqlStdOperatorTable.COALESCE,
+                operand0, operand0Zero),
+            rexBuilder.makeCall(SqlStdOperatorTable.COALESCE,
+                operand1, operand1Zero));
+
+        RexNode isNullCondition =
+            rexBuilder.makeCall(SqlStdOperatorTable.EQUALS,
+            rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL,
+                operand0),
+            rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL,
+                operand1));
+
+        newCall =
+            rexBuilder.makeCall(SqlStdOperatorTable.AND,
+                coalesceCondition, isNullCondition);
+      }
+      return newCall;
+    }
+  }
+
+  private static RexNode makeZeroLiteral(RexBuilder rexBuilder, RexNode operand) {
+    if (operand.getType().getSqlTypeName() == SqlTypeName.ARRAY) {
+      ArraySqlType arrayType = (ArraySqlType) operand.getType();
+      RelDataType elementType = arrayType.getComponentType();
+      RexNode elementZero = rexBuilder.makeZeroLiteral(elementType);
```

Comment:
You need to call the function recursively, because ARRAY can have ARRAY as elements.
Also, this may work for MULTISET too.

Reply:
there is also the case of ROW types.
Maybe you want to enhance the makeZeroLiteral function to handle these as well?

Reply:
Great Idea! I think I can create a new ticket to complete this suggestion first.

Reply:
@mihaibudiu I have used the new method `makeZeroRexNode` to replace the `makeZeroLiteral`. Could please help me to have a look again?